tools/mpremote/mpremote/transport_serial.py

```diff
@@ -95,29 +99,45 @@ def close(self):
         self.serial.close()
 
     def read_until(self, min_num_bytes, ending, timeout=10, data_consumer=None):
-        # if data_consumer is used then data is not accumulated and the ending must be 1 byte long
+        """
+        min_num_bytes: Return at least min_num_bytes.
+        ending: Return if 'ending' matches.
+        timeout [s]: Return if timeout between characters. None for infinite timeout.
+        data_consumer: Use callback for incoming characters.
+            If data_consumer is used then data is not accumulated and the ending must be 1 byte long
+
+        It is not visible for the caller why the function returned. It could be ending, timeout, min_num_bytes.
+        """
         assert data_consumer is None or len(ending) == 1
+        assert isinstance(timeout, (type(None), int, float))
 
-        data = self.serial.read(min_num_bytes)
-        if data_consumer:
-            data_consumer(data)
-        timeout_count = 0
+        data = b""
+        begin_char_s = time.monotonic()
         while True:
-            if data.endswith(ending):
+            if timeout and time.monotonic() > begin_char_s + timeout:
                 break
-            elif self.serial.inWaiting() > 0:
-                new_data = self.serial.read(1)
-                if data_consumer:
-                    data_consumer(new_data)
-                    data = new_data
-                else:
-                    data = data + new_data
-                timeout_count = 0
-            else:
-                timeout_count += 1
-                if timeout is not None and timeout_count >= 100 * timeout:
-                    break
+
+            if self.serial.inWaiting() <= 0:
+                # Wait for input
                 time.sleep(0.01)
+                continue
+
+            new_data = self.serial.read(1)  # Will not block!
+            begin_char_s = time.monotonic()
+            if data_consumer:
+                data_consumer(new_data)
+                data = new_data
+            else:
+                data = data + new_data
+
+            if min_num_bytes > 1:
+                # Make sure 'min_num_bytes' are read before
+                # testing for 'ending'.
```

Comment:
There's no need for this check, you can simply do `if data.endswith(ending)` (and it'll be `False` if there aren't yet enough bytes in `data`).

In fact, `min_num_bytes` is no longer needed with this PR.  That parameter was originally used for the initial blocking call to `read(min_num_bytes)` at the start of this function.